### PR TITLE
[PD]: Support HiCache prefetching and pd-incremental transfer on decode side

### PR DESCRIPTION
**Pull request by hzh0425**
## Motivation
Roadmap: #21846 、 https://github.com/sgl-project/sglang/issues/21703
Still in progress, TODO:

- [x] Precisely control token budget
- [x] Improve parallelization between HiCache transfer and PD transfer
- [ ] **Large-scale, systematic testing of PD + HiCache， Provide performance benefit metrics.**
- [ ] Support Unified Radix Tree

## Speed Tests and Profiling
Qwen32B + PD + HiCache + Mooncake

| Round | Prompt Length | No Decode Radix TTFT | Decode L2+L3 TTFT | TTFT Reduction | No Radix Hit Rate | L2+L3 Hit Rate | Hit Rate Gain |
|---:|---:|---:|---:|---:|---:|---:|---:|
| 0 | 4K | 2.29s | 1.59s | 30.9% | 0.000 | 0.000 | +0.000 |
| 1 | 16K | 7.25s | 5.56s | 23.3% | 0.250 | 0.750 | +0.500 |
| 2 | 28K | 12.72s | 9.19s | 27.8% | 0.571 | 0.857 | +0.286 |
| 3 | 40K | 18.01s | 11.93s | 33.8% | 0.700 | 0.900 | +0.200 |
| 4 | 52K | 23.40s | 12.52s | 46.5% | 0.769 | 0.923 | +0.154 |
| 5 | 64K | 28.50s | 15.18s | 46.7% | 0.812 | 0.937 | +0.125 |

| Metric | No Decode Radix | Decode L2+L3 | Change |
|---|---:|---:|---:|
| Average TTFT | 15.36s | 9.33s | -39.3% |
| P90 TTFT | 36.46s | 19.30s | -47.1% |
| P99 TTFT | 48.87s | 23.51s | -51.9% |
| Cache Hit Rate | 0.686 | 0.882 | +19.6 percentage points |
| Average TPOT / ITL | 40.31ms | 41.32ms | +2.5% |
| P90 TPOT / ITL | 52.62ms | 53.63ms | +1.9% |
| Output Throughput | 183.47 tok/s | 185.70 tok/s | +1.2% |
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #26766211033](https://github.com/sgl-project/sglang/actions/runs/26766211033)
Latest PR Test (Extra): :white_check_mark: [Run #26766210099](https://github.com/sgl-project/sglang/actions/runs/26766210099)

### CODE REVIEW

File: python/sglang/srt/disaggregation/decode.py [hunk 1]
```diff
@@ -885,20 +1017,31 @@ def pop_preallocated(
...
-            dst_kv_indices = self._pre_alloc(decode_req.req, prefix_indices, prefix_len)
+            dst_kv_indices = self._pre_alloc(
+                decode_req.req,
+                prefix_indices,
+                prefix_len,
+                total_prefix_len,
+            )
+            decode_req.prefix_match = prefix_match
+            if self.scheduler.enable_decode_hicache:
+                self._start_hicache_prefetch(decode_req.req, prefix_match)
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/decode.py:916]: Could this fail? Should we guard it in a try-catch?

File: python/sglang/srt/disaggregation/decode.py [hunk 2]
```diff
@@ -475,7 +536,72 @@ def _match_prefix_and_lock(self, req: Req) -> Tuple[torch.Tensor, int]:
...
+                    last_host_node.get_prefix_hash_values(last_host_node.parent)
+                    if self.tree_cache.hicache_storage_pass_prefix_keys
+                    else None
+                )
+                l3_storage_hit_length = self.tree_cache.query_storage_hit_length(
+                    last_host_node,
+                    suffix_tokens,
+                    last_hash,
+                    prefix_keys,
+                )
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/decode.py]: Will this diverge among the TP ranks?

File: python/sglang/srt/mem_cache/hiradix_cache.py [hunk 3]
```diff
@@ -1070,6 +1084,42 @@ def init_load_back(
...
 
+    def query_storage_hit_length(
+        self,
+        last_host_node: TreeNode,
+        new_input_tokens: List[int],
+        last_hash: Optional[str] = None,
+        prefix_keys: Optional[List[str]] = None,
+    ) -> int:
+        if not self.enable_storage or self.cache_controller.prefetch_rate_limited():
+            return 0
```
Comment (ShangmingCai) [inline python/sglang/srt/mem_cache/hiradix_cache.py:1095]: This branch seems to bypass `self._all_reduce_attn_groups`, not sure if it will cause tp rank divergence.

File: python/sglang/srt/disaggregation/decode.py (hunk 2 above)
Comment (hzh0425) [reply python/sglang/srt/disaggregation/decode.py]: query_storage_hit_length has already sync min tp

File: python/sglang/srt/mem_cache/hiradix_cache.py (hunk 3 above)
Comment (hzh0425) [reply python/sglang/srt/mem_cache/hiradix_cache.py:1095]: This is safe; its dependency, prefetch_occupied, is synchronized across ranks.

File: python/sglang/srt/disaggregation/decode.py (hunk 1 above)
Comment (hzh0425) [reply python/sglang/srt/disaggregation/decode.py:916]: Add a failback logic in here

**Comment by hzh0425** [comment]
/rerun-test test/registered/disaggregation/test_disaggregation_decode_radix_cache.py

**Comment by hzh0425** [comment]
/rerun-test test/registered/disaggregation/test_disaggregation_decode_radix_cache.py test/registered/disaggregation/test_disaggregation_pp.py

File: python/sglang/srt/disaggregation/decode.py [hunk 4]
```diff
@@ -1542,7 +1615,12 @@ def pop_transferred(self, rids_to_check: Optional[List[str]] = None) -> List[Req
                 if self.scheduler.metrics_reporter.enable_metrics:
                     self.scheduler.metrics_collector.increment_transfer_failed_reqs()
                 continue
-            elif poll == KVPoll.Success:
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/decode.py:1545]: Is there a reason why we change it from `elif` to `if`?

File: python/sglang/srt/disaggregation/decode.py [hunk 5]
```diff
@@ -1518,12 +1582,21 @@ def pop_transferred(self, rids_to_check: Optional[List[str]] = None) -> List[Req
...
+            ):
+                error_message = (
+                    f"Decode transfer failed for request rank={self.tp_rank} "
+                    f"{decode_req.req.rid=} {decode_req.req.bootstrap_room=}"
+                )
+                if poll == KVPoll.Failed:
+                    try:
+                        decode_req.kv_receiver.failure_exception()
+                    except Exception as e:
+                        error_message += f" with exception {e}"
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/decode.py:1598]: What could be the reason of `HiCacheRestoreResult.FAILED`? Should we append some error_message detail as well?

File: python/sglang/srt/disaggregation/decode_hicache_mixin.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+            cow_mamba=False,
+            include_req=True,
+        )
+        new_indices, restored_node = self.tree_cache.init_load_back(
+            InitLoadBackParams(
+                best_match_node=rematch.best_match_node,
+                host_hit_length=rematch.host_hit_length,
+                req=dr.req,
+            )
+        )
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/decode_hicache_mixin.py:211]: Will this mutate the tree node? If TP divergence happens, will the node mutation be cleaned in `_clean_hicache_prefetch_resources`?

**Review by ShangmingCai — APPROVED** [review]
No other comments.

File: python/sglang/srt/disaggregation/decode_hicache_mixin.py (hunk 6 above)
Comment (hzh0425) [reply python/sglang/srt/disaggregation/decode_hicache_mixin.py:211]: init_load_back triggers an L2 → L1 load, which typically does not fail.

File: python/sglang/srt/disaggregation/decode.py (hunk 5 above)
Comment (hzh0425) [reply python/sglang/srt/disaggregation/decode.py:1598]: Add a failed log in decode_hicache_mixin

File: python/sglang/srt/disaggregation/decode.py (hunk 4 above)
Comment (hzh0425) [reply python/sglang/srt/disaggregation/decode.py:1545]: change back

**Comment by hzh0425** [comment]
/rerun-test test/registered/disaggregation/test_disaggregation_decode_radix_cache.py

File: python/sglang/srt/mem_cache/hiradix_cache.py [hunk 7]
```diff
@@ -1070,6 +1084,42 @@ def init_load_back(
...
+            self.cache_controller.mem_pool_host.get_dummy_flat_data_page()[:0],
+            prefetch_key,
+            last_hash,
+            prefix_keys,
+        )
+        hash_values, storage_hit_count = self.cache_controller._storage_hit_query(
+            operation
+        )
+        storage_hit_count_tensor = torch.tensor(storage_hit_count, dtype=torch.int)
+        self._all_reduce_attn_groups(
```
Comment (ShangmingCai) [inline python/sglang/srt/mem_cache/hiradix_cache.py:1116]: We could revisit this in the future if this per-request `all_reduce` causes a performance issue for large batch sizes.

**Comment by ShangmingCai** [comment]
CI has passed.
image
image

**Comment by weibingo** [comment]
@hzh0425 I have a question: when the KV cache of the decode node is in L3, even if the cache hits in L3, it still needs to be pulled from L3 to the CPU, and then from the CPU to the GPU. The time spent in this process should be greater than the overhead of transferring from P to D, right?

**Comment by ShangmingCai** [comment]
> @hzh0425 I have a question: when the KV cache of the decode node is in L3, even if the cache hits in L3, it still needs to be pulled from L3 to the CPU, and then from the CPU to the GPU. The time spent in this process should be greater than the overhead of transferring from P to D, right?

@weibingo Yeah, but the matched prefix KV fetching could be overlapped with the computing and transferring of the incremental part, so this design is better for multi-turn conversations and agentic use cases.
